File: android-app/app/src/main/java/bypass/whitelist/discovery/VkDiscoveryScanner.kt
package bypass.whitelist.discovery

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Bitmap
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.view.ViewGroup
import android.webkit.CookieManager
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import bypass.whitelist.BuildConfig
import bypass.whitelist.tunnel.CallConfig
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.atomic.AtomicBoolean

object VkDiscoveryScanner {
    private const val GROUP_ID = "237416141"
    private const val UA = "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 Chrome/124 Mobile Safari/537.36"
    private val urls = listOf(
        "https://m.vk.ru/club$GROUP_ID",
        "https://m.vk.com/club$GROUP_ID",
        "https://vk.ru/club$GROUP_ID",
        "https://vk.com/club$GROUP_ID",
    )
    private val encryptedPayloadRegex = Regex("(wt(?:room|bus)2)\\.([A-Za-z0-9_-]{1,16})\\.([А-Яа-я]{24,})")
    private val payloadRegex = Regex("wt1\\.([A-Za-z0-9_-]{24,})")
    private val roomRegex = Regex("wbstream://[A-Za-z0-9._~:/?#\\[\\]@!$&'()*+,;=%-]+")

    data class Stats(
        val encryptedEvents: Int = 0,
        val legacyEvents: Int = 0,
        val slots: Int = 0,
        val free: Int = 0,
        val busy: Int = 0,
        val closed: Int = 0,
        val expired: Int = 0,
        val roomLinks: Int = 0,
    ) {
        fun summary(): String = "events=${encryptedEvents + legacyEvents} encrypted=$encryptedEvents legacy=$legacyEvents slots=$slots free=$free busy=$busy closed=$closed expired=$expired raw_rooms=$roomLinks"
    }

    data class Result(
        val configs: List<CallConfig>,
        val source: String?,
        val method: String = "http",
        val stats: Stats = Stats(),
    )

    private data class ParsedConfigs(val configs: List<CallConfig>, val stats: Stats)

    private data class Event(
        val slotId: String,
        val leaseId: String?,
        val status: String,
        val room: String?,
        val creator: String?,
        val node: String?,
        val location: String?,
        val createdAt: Long,
        val expiresAt: Long,
        val seq: Long,
    ) {
        val order: Long get() = if (seq > 0) seq else createdAt
    }

    fun sendClientEvent(
        type: String,
        clientId: String,
        room: String?,
        reason: String,
        badRooms: Collection<String> = emptyList(),
    ): Boolean {
        val token = BuildConfig.VK_BOT_TOKEN.takeIf { it.isNotBlank() } ?: return false
        val peerId = BuildConfig.VK_BOT_PEER_ID.takeIf { it.isNotBlank() } ?: return false
        return try {
            val now = System.currentTimeMillis() / 1000L
            val payload = JSONObject().apply {
                put("v", 2)
                put("type", type)
                put("client_id", clientId)
                put("platform", "android")
                put("app_version", BuildConfig.VERSION_NAME)
                put("app_build", BuildConfig.VERSION_CODE)
                put("device", "${Build.MANUFACTURER} ${Build.MODEL}".trim())
                put("system_version", Build.VERSION.RELEASE ?: "")
                put("room", room ?: "")
                put("bad_rooms", JSONArray().also { arr -> badRooms.forEach { arr.put(it) } })
                put("reason", reason)
                put("created_at", now)
                put("seq", now)
                put("nonce", java.util.UUID.randomUUID().toString())
            }
            val message = WtBusCrypto.encryptEnvelope("wtclient2", payload) ?: return false
            val body = listOf(
                "peer_id" to peerId,
                "random_id" to System.currentTimeMillis().toString(),
                "message" to message,
                "access_token" to token,
                "v" to "5.199",
            ).joinToString("&") { (k, v) -> "${k}=${URLEncoder.encode(v, "UTF-8")}" }
            val conn = URL("https://api.vk.com/method/messages.send").openConnection() as HttpURLConnection
            conn.instanceFollowRedirects = true
            conn.requestMethod = "POST"
            conn.connectTimeout = 8000
            conn.readTimeout = 10000
            conn.doOutput = true
            conn.setRequestProperty("User-Agent", "BEZabotny-NET Android private bus")
            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val json = JSONObject(conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() })
            !json.has("error") && json.has("response")
        } catch (_: Exception) {
            false
        }
    }

    fun scan(): Result {
        val privateHtml = fetchPrivateBus()
        if (privateHtml != null) {
            val configs = parseConfigs(privateHtml)
            if (configs.isNotEmpty()) return Result(configs, "vk-private-bus", "vk-private")
        }
        var source: String? = null
        for (url in urls) {
            val html = fetch(url) ?: continue
            source = url
            val configs = parseConfigs(html)
            if (configs.isNotEmpty()) return Result(configs, source, "http")
        }
        return Result(emptyList(), source, "http")
    }


    private fun fetchPrivateBus(): String? {
        val token = BuildConfig.VK_BOT_TOKEN.takeIf { it.isNotBlank() } ?: return null
        val peerId = BuildConfig.VK_BOT_PEER_ID.takeIf { it.isNotBlank() } ?: return null
        return try {
            val body = listOf(
                "peer_id" to peerId,
                "count" to "100",
                "access_token" to token,
                "v" to "5.199",
            ).joinToString("&") { (k, v) -> "${k}=${URLEncoder.encode(v, "UTF-8")}" }
            val conn = URL("https://api.vk.com/method/messages.getHistory").openConnection() as HttpURLConnection
            conn.instanceFollowRedirects = true
            conn.requestMethod = "POST"
            conn.connectTimeout = 8000
            conn.readTimeout = 10000
            conn.doOutput = true
            conn.setRequestProperty("User-Agent", "BEZabotny-NET private bus")
            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val json = JSONObject(conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() })
            val items = json.optJSONObject("response")?.optJSONArray("items") ?: return null
            buildString {
                for (i in 0 until items.length()) {
                    val text = items.optJSONObject(i)?.optString("text").orEmpty()
                    if (text.isNotBlank()) append(text).append("\n")
                }
            }
        } catch (_: Exception) {
            null
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    fun scanWithWebView(activity: Activity, onProgress: (String) -> Unit = {}, onDone: (Result) -> Unit) {
        val httpResult = scan()
        if (httpResult.configs.isNotEmpty()) {
            onDone(httpResult)
            return
        }
        val done = AtomicBoolean(false)
        val handler = Handler(Looper.getMainLooper())
        val webView = WebView(activity)
        fun cleanup() {
            try { (webView.parent as? ViewGroup)?.removeView(webView) } catch (_: Exception) {}
            try { webView.stopLoading() } catch (_: Exception) {}
            try { webView.destroy() } catch (_: Exception) {}
        }
        fun finish(result: Result) {
            if (!done.compareAndSet(false, true)) return
            cleanup()
            onDone(result)
        }
        var index = 0
        fun loadNext() {
            if (done.get()) return
            if (index >= urls.size) {
                finish(Result(emptyList(), null, "webview"))
                return
            }
            val url = urls[index++]
            onProgress("WebView VK: ${url.substringAfter("://").substringBefore('/')}")
            webView.loadUrl(url)
        }
        fun inspect(url: String?) {
            if (done.get()) return
            val js = """
                (function(){
                  var h=document.documentElement?document.documentElement.outerHTML:'';
                  var t=document.body?document.body.innerText:'';
                  return h+'\n'+t;
                })();
            """.trimIndent()
            webView.evaluateJavascript(js) { encoded ->
                if (done.get()) return@evaluateJavascript
                val raw = decodeJsString(encoded)
                val parsed = parseConfigsDetailed(raw)
                if (parsed.configs.isNotEmpty()) {
                    onProgress("VK page parsed: ${parsed.configs.size} free room(s); ${parsed.stats.summary()}")
                    finish(Result(parsed.configs, url ?: webView.url, "webview", parsed.stats))
                } else {
                    handler.postDelayed({ loadNext() }, 500)
                }
            }
        }
        val timeout = Runnable { finish(Result(emptyList(), webView.url, "webview-timeout", Stats())) }
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        webView.settings.cacheMode = WebSettings.LOAD_DEFAULT
        webView.settings.userAgentString = UA
        CookieManager.getInstance().setAcceptCookie(true)
        CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true)
        webView.webChromeClient = WebChromeClient()
        webView.webViewClient = object : WebViewClient() {
            override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? = null
            override fun onPageStarted(view: WebView, url: String?, favicon: Bitmap?) { onProgress("VK page loading…") }
            override fun onPageFinished(view: WebView, url: String?) { handler.postDelayed({ inspect(url) }, 2500) }
        }
        handler.postDelayed(timeout, 30000)
        loadNext()
    }

    fun parseConfigs(raw: String): List<CallConfig> = parseConfigsDetailed(raw).configs

    private fun parseConfigsDetailed(raw: String): ParsedConfigs {
        val html = normalize(raw)
        val encryptedEvents = encryptedPayloadRegex.findAll(html).mapNotNull { match ->
            WtBusCrypto.decryptEnvelope(match.groupValues[1], match.groupValues[2], match.groupValues[3])?.toEvent()
        }.toList()
        val legacyEvents = payloadRegex.findAll(html).mapNotNull { decodeJson(it.groupValues[1])?.toEvent() }.toList()
        val events = encryptedEvents + legacyEvents
        val roomLinks = roomRegex.findAll(html).map { it.value }.distinct().toList()
        if (events.isNotEmpty()) {
            val newestBySlot = events.groupBy { it.slotId }.mapValues { (_, list) -> list.maxBy { it.order } }
            val now = System.currentTimeMillis() / 1000L
            val newest = newestBySlot.values
            val stats = Stats(
                encryptedEvents = encryptedEvents.size,
                legacyEvents = legacyEvents.size,
                slots = newestBySlot.size,
                free = newest.count { it.status == "free" && it.room?.startsWith("wbstream://") == true && it.expiresAt > now },
                busy = newest.count { it.status == "busy" },
                closed = newest.count { it.status == "closed" },
                expired = newest.count { it.expiresAt <= now },
                roomLinks = roomLinks.size,
            )
            val configs = newest
                .filter { it.status == "free" && it.expiresAt > now && it.room?.startsWith("wbstream://") == true }
                .sortedByDescending { it.order }
                .map { event ->
                    val label = listOfNotNull(event.node?.ifBlank { null } ?: event.creator?.ifBlank { null }, event.location?.ifBlank { null }).joinToString(" · ").ifBlank { null }
                    CallConfig.autoWith(
                        name = label ?: event.creator?.ifBlank { null } ?: "VK discovery",
                        url = event.room!!,
                        slotId = event.slotId,
                        leaseId = event.leaseId,
                        expiresAt = event.expiresAt,
                        nodeLabel = label,
                    )
                }
            return ParsedConfigs(configs, stats)
        }
        val configs = roomLinks.map { room ->
            CallConfig.autoWith(name = "VK discovery", url = room, slotId = room, leaseId = null, expiresAt = null)
        }
        return ParsedConfigs(configs, Stats(roomLinks = roomLinks.size, free = configs.size))
    }

    private fun JSONObject.toEvent(): Event? {
        val status = optString("status").ifBlank { "free" }
        val room = optString("room").takeIf { it.isNotBlank() && it != "null" }
        val streamId = optString("stream_id").takeIf { it.isNotBlank() }
        val slotId = optString("slot_id").takeIf { it.isNotBlank() } ?: streamId ?: room ?: return null
        val leaseId = optString("lease_id").takeIf { it.isNotBlank() } ?: streamId
        val createdAt = optLong("created_at", 0L)
        val expiresAt = optLong("expires_at", Long.MAX_VALUE / 4)
        val seq = optLong("seq", 0L)
        return Event(
            slotId = slotId,
            leaseId = leaseId,
            status = status,
            room = room,
            creator = optString("creator").takeIf { it.isNotBlank() },
            node = optString("node").takeIf { it.isNotBlank() } ?: optString("node_name").takeIf { it.isNotBlank() },
            location = optString("location").takeIf { it.isNotBlank() }
                ?: listOfNotNull(optString("country").takeIf { it.isNotBlank() }, optString("city").takeIf { it.isNotBlank() }, optString("region").takeIf { it.isNotBlank() }).joinToString(" ").takeIf { it.isNotBlank() },
            createdAt = createdAt,
            expiresAt = expiresAt,
            seq = seq,
        )
    }

    private fun normalize(raw: String): String {
        return raw
            .replace("\\u002F", "/")
            .replace("\\/", "/")
            .replace("&quot;", "\"")
            .replace("&#34;", "\"")
            .replace("&amp;", "&")
    }

    private fun decodeJson(encoded: String): JSONObject? {
        return try {
            val padded = encoded + "=".repeat((4 - encoded.length % 4) % 4)
            val bytes = Base64.decode(padded, Base64.URL_SAFE or Base64.NO_WRAP)
            JSONObject(String(bytes, Charsets.UTF_8))
        } catch (_: Exception) {
            null
        }
    }

    private fun decodeJsString(value: String?): String {
        if (value == null || value == "null") return ""
        return try {
            JSONObject("{\"v\":$value}").optString("v")
        } catch (_: Exception) {
            value.trim('"')
                .replace("\\n", "\n")
                .replace("\\u003C", "<")
                .replace("\\\"", "\"")
        }
    }

    private fun fetch(url: String): String? {
        return try {
            val conn = URL(url).openConnection() as HttpURLConnection
            conn.instanceFollowRedirects = true
            conn.connectTimeout = 8000
            conn.readTimeout = 10000
            conn.setRequestProperty("User-Agent", UA)
            conn.setRequestProperty("Accept", "text/html,application/xhtml+xml")
            conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        } catch (_: Exception) {
            null
        }
    }
}

File: ios-proxy-app/whitelist-bypass-proxy/VKDiscoveryScanner.swift
import Foundation
import os
import UIKit

struct DiscoveryRoom: Identifiable {
    let id: String
    let status: String
    let room: String
    let creator: String
    let node: String
    let location: String
    let createdAt: Int64
    let expiresAt: Int64
    let seq: Int64

    var order: Int64 { seq > 0 ? seq : createdAt }
    var isFree: Bool {
        status == "free" && !room.isEmpty && expiresAt > Int64(Date().timeIntervalSince1970)
    }
    var displayName: String {
        let left = node.isEmpty ? creator : node
        if !left.isEmpty && !location.isEmpty { return "\(left) · \(location)" }
        if !left.isEmpty { return left }
        return "VK discovery"
    }
}

final class VKDiscoveryScanner {
    private let logger = Logger(subsystem: "bypass.whitelist", category: "discovery")
    private let groupId = "237416141"
    private let userAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 Version/17.0 Mobile/15E148 Safari/604.1"
    private var urls: [URL] {
        [
            "https://m.vk.ru/club\(groupId)",
            "https://m.vk.com/club\(groupId)",
            "https://vk.ru/club\(groupId)",
            "https://vk.com/club\(groupId)",
        ].compactMap(URL.init(string:))
    }

    func scan(completion: @escaping (_ rooms: [DiscoveryRoom], _ source: String?) -> Void) {
        logger.info("scanner scan started")
        scanPrivateBus { [weak self] rooms in
            guard let self = self else { return }
            if !rooms.isEmpty {
                self.logger.info("private-bus returned rooms=\(rooms.count, privacy: .public) free=\(rooms.filter { $0.isFree }.count, privacy: .public)")
                completion(rooms, "vk-private-bus")
                return
            }
            self.logger.warning("private-bus returned no rooms; falling back to wall")
            self.scanWall(urls: self.urls, firstSource: nil, completion: completion)
        }
    }


    func sendClientEvent(type: String, clientId: String, room: String?, reason: String, badRooms: [String] = [], completion: ((Bool) -> Void)? = nil) {
        guard !WtBusSecrets.vkBotToken.isEmpty, !WtBusSecrets.vkBotPeerID.isEmpty else {
            logger.warning("client event skipped: empty token or peer id")
            completion?(false)
            return
        }
        let now = Int(Date().timeIntervalSince1970)
        let payload: [String: Any] = [
            "v": 2,
            "type": type,
            "client_id": clientId,
            "platform": "ios",
            "app_version": AppVersion.name,
            "app_build": AppVersion.code,
            "device": UIDevice.current.model,
            "system_version": UIDevice.current.systemVersion,
            "room": room ?? "",
            "bad_rooms": badRooms,
            "reason": reason,
            "created_at": now,
            "seq": now,
            "nonce": UUID().uuidString,
        ]
        guard let message = WtBusCrypto.encryptEnvelope(prefix: "wtclient2", payload: payload),
              let url = URL(string: "https://api.vk.com/method/messages.send") else {
            logger.error("client event encryption failed type=\(type, privacy: .public)")
            completion?(false)
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 10
        request.setValue("BEZabotny-NET iOS private bus", forHTTPHeaderField: "User-Agent")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let params: [(String, String)] = [
            ("peer_id", WtBusSecrets.vkBotPeerID),
            ("random_id", String(Int(Date().timeIntervalSince1970 * 1000))),
            ("message", message),
            ("access_token", WtBusSecrets.vkBotToken),
            ("v", "5.199"),
        ]
        request.httpBody = params.map { key, value in
            "\(key)=\(value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value)"
        }.joined(separator: "&").data(using: .utf8)
        URLSession.shared.dataTask(with: request) { [weak self] data, _, error in
            if let error = error {
                self?.logger.error("client event send failed type=\(type, privacy: .public) error=\(error.localizedDescription, privacy: .public)")
                completion?(false)
                return
            }
            let ok: Bool
            if let data = data,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                ok = json["error"] == nil && json["response"] != nil
            } else {
                ok = false
            }
            self?.logger.info("client event sent type=\(type, privacy: .public) ok=\(ok, privacy: .public)")
            completion?(ok)
        }.resume()
    }

    private func scanPrivateBus(completion: @escaping ([DiscoveryRoom]) -> Void) {
        guard !WtBusSecrets.vkBotToken.isEmpty, !WtBusSecrets.vkBotPeerID.isEmpty else {
            logger.warning("private-bus skipped: empty token or peer id")
            completion([])
            return
        }
        guard let url = URL(string: "https://api.vk.com/method/messages.getHistory") else {
            logger.error("private-bus URL construction failed")
            completion([])
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 10
        request.setValue("BEZabotny-NET iOS private bus", forHTTPHeaderField: "User-Agent")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let params: [(String, String)] = [
            ("peer_id", WtBusSecrets.vkBotPeerID),
            ("count", "100"),
            ("access_token", WtBusSecrets.vkBotToken),
            ("v", "5.199"),
        ]
        request.httpBody = params.map { key, value in
            "\(key)=\(value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value)"
        }.joined(separator: "&").data(using: .utf8)
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }
            if let error = error {
                self.logger.error("private-bus network error: \(error.localizedDescription, privacy: .public)")
                completion([])
                return
            }
            if let http = response as? HTTPURLResponse {
                self.logger.info("private-bus HTTP \(http.statusCode, privacy: .public)")
            }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let response = json["response"] as? [String: Any],
                  let items = response["items"] as? [[String: Any]] else {
                self.logger.error("private-bus parse failed bytes=\(data?.count ?? 0, privacy: .public)")
                completion([])
                return
            }
            let text = items.compactMap { $0["text"] as? String }.joined(separator: "\n")
            let rooms = self.parse(text: text)
            self.logger.info("private-bus messages=\(items.count, privacy: .public) rooms=\(rooms.count, privacy: .public)")
            completion(rooms)
        }.resume()
    }

    private func scanWall(urls: [URL], firstSource: String?, completion: @escaping ([DiscoveryRoom], String?) -> Void) {
        guard let url = urls.first else {
            completion([], firstSource)
            return
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 10
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("text/html,application/xhtml+xml", forHTTPHeaderField: "Accept")
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }
            let source = firstSource ?? url.absoluteString
            if let error = error {
                self.logger.warning("wall fetch failed url=\(url.absoluteString, privacy: .public) error=\(error.localizedDescription, privacy: .public)")
            } else if let http = response as? HTTPURLResponse {
                self.logger.info("wall fetch HTTP \(http.statusCode, privacy: .public) url=\(url.absoluteString, privacy: .public)")
            }
            if let data = data, let html = String(data: data, encoding: .utf8) {
                let rooms = self.parse(text: html)
                self.logger.info("wall parse url=\(url.absoluteString, privacy: .public) bytes=\(data.count, privacy: .public) rooms=\(rooms.count, privacy: .public)")
                if !rooms.isEmpty {
                    completion(rooms, source)
                    return
                }
            }
            self.scanWall(urls: Array(urls.dropFirst()), firstSource: source, completion: completion)
        }.resume()
    }

    private func parse(text raw: String) -> [DiscoveryRoom] {
        let text = normalize(raw)
        var rooms: [DiscoveryRoom] = []
        rooms.append(contentsOf: parseEncryptedPayloads(text))
        rooms.append(contentsOf: parseLegacyPayloads(text))
        rooms.append(contentsOf: parseLegacyRooms(text))
        var unique: [String: DiscoveryRoom] = [:]
        for room in rooms { unique[room.id] = room }
        let sorted = unique.values.sorted { $0.order > $1.order }
        logger.info("parse summary encrypted=\(rooms.count, privacy: .public) unique=\(sorted.count, privacy: .public) free=\(sorted.filter { $0.isFree }.count, privacy: .public)")
        return sorted
    }

    private func parseEncryptedPayloads(_ text: String) -> [DiscoveryRoom] {
        guard let regex = try? NSRegularExpression(pattern: "(wt(?:room|bus)2)\\.([A-Za-z0-9_-]{1,16})\\.([А-Яа-я]{24,})") else { return [] }
        let ns = text as NSString
        return regex.matches(in: text, range: NSRange(location: 0, length: ns.length)).compactMap { match in
            guard match.numberOfRanges >= 4 else { return nil }
            let prefix = ns.substring(with: match.range(at: 1))
            let kid = ns.substring(with: match.range(at: 2))
            let encoded = ns.substring(with: match.range(at: 3))
            guard let json = WtBusCrypto.decryptEnvelope(prefix: prefix, kid: kid, encoded: encoded) else { return nil }
            return jsonToRoom(json)
        }
    }

    private func parseLegacyPayloads(_ text: String) -> [DiscoveryRoom] {
        guard let regex = try? NSRegularExpression(pattern: "wt1\\.([A-Za-z0-9_-]{24,})") else { return [] }
        let ns = text as NSString
        return regex.matches(in: text, range: NSRange(location: 0, length: ns.length)).compactMap { match in
            guard match.numberOfRanges >= 2 else { return nil }
            let encoded = ns.substring(with: match.range(at: 1))
            guard let data = base64URLDecode(encoded),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
            return jsonToRoom(json)
        }
    }

    private func jsonToRoom(_ json: [String: Any]) -> DiscoveryRoom {
        let status = (json["status"] as? String) ?? "free"
        let room = (json["room"] as? String) ?? ""
        let creator = (json["creator"] as? String) ?? ""
        let node = (json["node"] as? String) ?? (json["node_name"] as? String) ?? ""
        let location = (json["location"] as? String) ?? [json["country"] as? String, json["city"] as? String, json["region"] as? String].compactMap { $0 }.joined(separator: " ")
        let createdAt = int64(json["created_at"]) ?? int64(json["createdAt"]) ?? 0
        let expiresAt = int64(json["expires_at"]) ?? int64(json["expiresAt"]) ?? Int64.max
        let seq = int64(json["seq"]) ?? 0
        let streamId = (json["stream_id"] as? String) ?? (json["streamId"] as? String) ?? room
        return DiscoveryRoom(id: streamId.isEmpty ? room : streamId, status: status, room: room, creator: creator, node: node, location: location, createdAt: createdAt, expiresAt: expiresAt, seq: seq)
    }

    private func parseLegacyRooms(_ text: String) -> [DiscoveryRoom] {
        guard let regex = try? NSRegularExpression(pattern: "wbstream://[A-Za-z0-9._~:/?#\\[\\]@!$&'()*+,;=%-]+") else { return [] }
        let ns = text as NSString
        return regex.matches(in: text, range: NSRange(location: 0, length: ns.length)).map { match in
            let room = ns.substring(with: match.range)
            return DiscoveryRoom(id: room, status: "free", room: room, creator: "legacy", node: "legacy", location: "", createdAt: 0, expiresAt: Int64.max, seq: 0)
        }
    }

    private func normalize(_ raw: String) -> String {
        raw
            .replacingOccurrences(of: "\\u002F", with: "/")
            .replacingOccurrences(of: "\\/", with: "/")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#34;", with: "\"")
            .replacingOccurrences(of: "&amp;", with: "&")
    }

    private func int64(_ value: Any?) -> Int64? {
        if let n = value as? Int64 { return n }
        if let n = value as? Int { return Int64(n) }
        if let n = value as? Double { return Int64(n) }
        if let s = value as? String { return Int64(s) }
        return nil
    }

    private func base64URLDecode(_ text: String) -> Data? {
        var s = text.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let pad = s.count % 4
        if pad > 0 { s += String(repeating: "=", count: 4 - pad) }
        return Data(base64Encoded: s)
    }
}
